buildflow/runtime/ray_io/sqs_io.py

```diff
@@ -0,0 +1,136 @@
+
+    def backlog(self) -> Optional[float]:
+        client = self.get_boto_client()
+        queue_url = self.get_queue_url()
+        queue_atts = client.get_queue_attributes(
+            QueueUrl=queue_url, AttributeNames=['ApproximateNumberOfMessages'])
+        if 'ApproximateNumberOfMessages' in queue_atts['Attributes']:
+            return int(queue_atts['Attributes']['ApproximateNumberOfMessages'])
+        return 0
+
+    @classmethod
+    def recommended_num_threads(cls):
+        # TODO: tune this value.
+        return 4
+
+    @classmethod
+    def num_cpus(cls) -> float:
+        return .25
+
+
+@ray.remote(num_cpus=SQSSource.num_cpus())
+class SQSSourceActor(base.StreamingRaySource):
+
+    def __init__(self, ray_sinks: Dict[str, base.RaySink],
+                 source: SQSSource) -> None:
+        super().__init__(ray_sinks)
+        if source._test_sqs_client is not None:
+            self.sqs_client = source._test_sqs_client
+        else:
+            self.sqs_client = boto3.client('sqs')
+        self.queue_url = _get_queue_url(self.sqs_client, source.queue_name,
+                                        source.queue_owner_aws_account_id)
+        # This is the max messages allowed by SQS.
+        self.batch_size = 10
+        self.running = True
+
+    async def run(self):
+        while self.running:
+            try:
+                response = self.sqs_client.receive_message(
+                    QueueUrl=self.queue_url,
+                    AttributeNames=['All'],
+                    MaxNumberOfMessages=self.batch_size)
+                num_messages = 0
+                if 'Messages' in response:
+                    num_messages = len(response['Messages'])
+                    await self._send_batch_to_sinks_and_await(
+                        response['Messages'])
+                    to_delete = []
+                    for message in response['Messages']:
+                        to_delete.append({
+                            'Id':
+                            message['MessageId'],
+                            'ReceiptHandle':
+                            message['ReceiptHandle']
+                        })
+                    self.sqs_client.delete_message_batch(
+                        QueueUrl=self.queue_url, Entries=to_delete)
+                # TODO: we should look into abstracting the while loop in
+                # base.StreamingRaySource then new sources wouldn't have to
+                # worry about the shutdown / reporting metrics
+                self.update_metrics(num_messages)
+                # Since SQS doesn't have an async client we need to
+                # sleep here to yield back the event loop. This allows us to
+                # collect metrics and shut down correctly.
+                # Ideally we wouldn't have to put this sleep in here though.
+                # TODO: look at using promethius metrics to get metrics
+                # instead.
+                await asyncio.sleep(1)
```

Comment:
If its just for releasing the event loop, what if we sleep for 0.1?

Reply:
done.